### PR TITLE
chore: remove max length from pytorch trials

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

remove max_length from pytorch trials/trainer

## Checklist

- [ ] Changes have been manually QA'd
- [ ] New features have been approved by the corresponding PM
- [ ] User-facing API changes have the "User-facing API Change" label
- [ ] Release notes have been added as a separate file under `docs/release-notes/`
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses have been included for new code which was copied and/or modified from any external code

[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: harness/determined/pytorch/_pytorch_trial.py
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:123]: nit: `f"Epoch({self.value})"`; better to render it as the user is going to type it out.

Same with Batch.

File: examples/tutorials/mnist_pytorch/train.py
Comment (rb-determined-ai) [inline examples/tutorials/mnist_pytorch/train.py:83]: I actually like making the user emit the special metrics they care about.  It is not difficult.

If we are going to do it, I feel pretty strongly that we need to stick to `steps_completed` as the name for batches, since that's basically everywhere in our system, and I don't want to introduce `batch` as an off-by-one alias of `steps_completed` soley for the purpose of having `time_unit: batch` look nicer than `time_unit: steps_completed`.

I'm open to either option.

File: harness/determined/pytorch/_pytorch_trial.py
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:528]: I don't care for this trying to guess which special metric to add.

I think, either we add `steps_completed` and `epoch` to validation metrics always, or we do it never.  I don't see why max_length should affect that.

File: harness/determined/pytorch/_trainer.py
Comment (rb-determined-ai) [inline harness/determined/pytorch/_trainer.py:1]: Docstring update required:

```
            max_length: The maximum number of steps to train for. This value is required and
                only applicable in local training mode. For on-cluster training, this value will
                be ignored; the searcher’s ``max_length`` must be configured from the experiment
                configuration. This is a ``TrainUnit`` type (``Batch`` or ``Epoch``) which takes an
``int``. For example, ``Epoch(1)`` would train for a maximum length of one epoch.
```

File: harness/determined/pytorch/_trainer.py
Comment (rb-determined-ai) [inline harness/determined/pytorch/_trainer.py:188]: Please indicate which version it was deprecated in.  I've been using the `XXYYZZ` placeholder for the version, which I'll go through and update before we land it.

**Review by rb-determined-ai — COMMENTED** [review]
Looks good, it's basically there already.  But I do want to do a final review before landing it.

File: examples/tutorials/mnist_pytorch/train.py
Comment (azhou-determined) [reply examples/tutorials/mnist_pytorch/train.py:83]: my reasoning for not making the user do it:
- this is a managed training loop, user isn't calling `core.train.report_metrics` themselves. there's some expectation that it'll "just work".
- we don't support anything besides "batch"/"epoch" for `time_metric` in PyTorchTrial/Trainer anyway. so it's not that the user cares about these "batch"/"epoch" metrics, it's that we do. 

my reasoning for making the user do it:
- >It is not difficult.
- we already make them do this if they want "epoch" behavior in the web UI. 
- the validation metrics aren't arbitrary as is, you must report the searcher metric yourself which is also configured in the expconf, and this feels similar. 
- it's explicit about what the system actually expects, no magic.

i can see both ways as well, but "no magic" is pretty compelling to me, and i _really_ don't want to do `time_unit: steps_completed`, so i changed it back to "user must report" behavior.

Comment (rb-determined-ai) [reply examples/tutorials/mnist_pytorch/train.py:83]: We talked out-of-band, and agreed that:
- We like the simplicity of "you must emit the time_metric" you want for the searcher to work
- We also like the ease-of-use of just reporting "epochs" and "batches" automatically, which additionally supports the epochs-axis in the webui (with some small backend changes)
- We don't like magically including extra metrics based on what the time_metric is.  "epochs" and "batches" should just be emitted unconditionally.

File: harness/determined/pytorch/_pytorch_trial.py
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:253]: It occurs to me, seeing this case no longer supported here, that you need to add a `max_length` to the pytorch code in `exec/harness.py`, and you need to probably have this exact check over there.

File: harness/determined/pytorch/_trainer.py
Comment (rb-determined-ai) [inline harness/determined/pytorch/_trainer.py:105]: shouldn't this say `time_metric: "epochs'`?

File: harness/determined/pytorch/_trainer.py
Comment (rb-determined-ai) [inline harness/determined/pytorch/_trainer.py:205]: Can you say `max_length must either be a Batch() or Epoch() type` instead?

Not sure that the `TrainUnit` abstract type will make immediate sense to readers.

File: harness/determined/pytorch/_trainer.py
Comment (rb-determined-ai) [inline harness/determined/pytorch/_trainer.py:262]: 'batches' / 'epochs'?

Probably in other places I did not see, also.

File: harness/determined/pytorch/_pytorch_trial.py
Comment (rb-determined-ai) [reply harness/determined/pytorch/_pytorch_trial.py:253]: no... nvm, I think you have it correct.

I think you need the shim to be Trainer, not in harness.py, to cover all the cases.

File: harness/determined/pytorch/_trainer.py
Comment (rb-determined-ai) [inline harness/determined/pytorch/_trainer.py:191]: please use `warnings.warning()` for deprecation warnings, grep around for examples

File: harness/determined/pytorch/_trainer.py
Comment (rb-determined-ai) [inline harness/determined/pytorch/_trainer.py:157]: random question, though not necessarily part of this change:

Wouldn't it be better to do: `if not isinstance(max_length, (Batch, Epoch))`?  The error message is like you're testing `max_length` but the check is actually applying to `max_length.value`.

Also below in the backwards compatibility checks.

**Review by rb-determined-ai — APPROVED** [review]
excellent!

File: harness/determined/pytorch/_trainer.py
Comment (azhou-determined) [reply harness/determined/pytorch/_trainer.py]: sort of, there's two separate checks here that i'm confusing. 1 that max length is of Batch/Epoch class, 2 that what was passed into Batch/Epoch is an int (and not a collection, which we support for validation/checkpoint periods). 

so i combined these two checks into a combined error message.

**Comment by azhou-determined** [comment]
merged this in new branch https://github.com/determined-ai/determined/pull/10091 b/c of messy rebase

**Pull request closed without merging**
